src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+    let e = maybeIntern(this.record.value);
+    for(let field in this.fields) {
+      for(let dslValue of this.fields[field]) {
+        let value = this.block.toValue(dslValue) as (RawValue | Register);
+        scans.push(new Scan(e, maybeIntern(field), maybeIntern(value), IGNORE_REG))
+      }
+    }
+    return scans;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLBlock
+//--------------------------------------------------------------------
+
+class DSLBlock {
+  records:DSLRecord[] = [];
+  variables:DSLVariable[] = [];
+  functions:DSLFunction[] = [];
+  variableLookup:{[name:number]:DSLVariable[]} = {};
+  block:runtime.Block;
+
+  constructor(public name:string, public creationFunction:string) {
+    let functionArgs:string[] = [];
+    let code = creationFunction.toString();
+    // trim the function(...) { from the start and capture the arg names
+    code = code.replace(/function\s*\((.*)\)\s*\{/, function(str:string, args:string) {
+      functionArgs.push.apply(functionArgs, args.split(",").map((str) => str.trim()));
+      return "";
+    });
+    // trim the final } since we removed the function bit
+    code = code.substring(0, code.length - 1);
+    code = this.transformBlockCode(code, functionArgs);
+    let neueFunc = new Function(functionArgs[0], functionArgs[1], functionArgs[2], code);
+    neueFunc(this.find, this.record, this.generateLib());
+    this.prepare();
+  }
+
+  generateLib() {
+    let fnGet = (obj:any, prop:string) => {
+      let path = obj.path || [];
+      path.push(prop);
+      let neue:any = () => {};
+      neue.path = path;
+      return new Proxy(neue, {
+        get: fnGet,
+        apply: (target:any, targetThis:any, args:any[]) => {
+          let func = new DSLFunction(this, path, args);
+          this.functions.push(func);
+          return func;
+        }});
+    }
+    return new Proxy({}, {get:fnGet});
+  }
+
+  find = (...args:any[]) => {
+    let lastArg = args[args.length - 1];
+    let proxied:any = {};
+    let tag;
+    if(typeof lastArg === "object") {
+      proxied = lastArg;
+      tag = args.slice(0, args.length - 1);
+    } else {
+      tag = args.slice(0, args.length);
+    }
+    // console.log("tag", tag);
+    let rec = new DSLRecord(this, tag, proxied);
+    this.records.push(rec);
+    return rec.proxy();
+  }
+
+  record = (...args:any[]) => {
+    let out = this.find.apply(null, args);
+    out.output = true;
+    return out;
+  }
+
+  registerVariable(variable:DSLVariable) {
+    this.variableLookup[variable.id] = [variable];
+  }
+
+  toValue(a:any) {
+    if(a === undefined || a === null) throw new Error("Eve values can't be undefined or null");
+    if(a instanceof DSLVariable) {
+      return a.value;
+    } if(a instanceof DSLRecord) {
+      return a.record.value;
+    } if(a instanceof DSLFunction) {
+      return a.returnValue.value;
+    }
+    return a;
+  }
+
+  equivalence(a:any, b:any) {
```

Comment:
Let's spare a couple comments talking about the cases here.

Each value is either:
 - a static (interned) eve value
 - a record
 - a register (variable)

In the first case direct comparison is sufficient to test equality. In the second, etc. etc.